// durable-functions/lib/src/durableorchestrationcontext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class DurableOrchestrationContext {
    callActivity(name, input) {
        throw new Error("This is a placeholder.");
    }
    callActivityWithRetry(name, retryOptions, input) {
        throw new Error("This is a placeholder.");
    }
    callEntity(entityId, operationName, operationInput) {
        throw new Error("This is a placeholder.");
    }
    callSubOrchestrator(name, input, instanceId) {
        throw new Error("This is a placeholder.");
    }
    callSubOrchestratorWithRetry(name, retryOptions, input, instanceId) {
        throw new Error("This is a placeholder.");
    }
    callHttp(method, uri, content, headers, tokenSource) {
        throw new Error("This is a placeholder");
    }
    continueAsNew(input) {
        throw new Error("This is a placeholder.");
    }
    createTimer(fireAt) {
        throw new Error("This is a placeholder.");
    }
    getInput() {
        throw new Error("This is a placeholder.");
    }
    newGuid() {
        throw new Error("This is a placeholder.");
    }
    setCustomStatus(customStatusObject) {
        throw new Error("This is a placeholder.");
    }
    waitForExternalEvent(name) {
        throw new Error("This is a placeholder.");
    }
}
exports.DurableOrchestrationContext = DurableOrchestrationContext;
