// durable-functions/lib/src/tasks/taskfactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const task_1 = require("./task");
const taskset_1 = require("./taskset");
const timertask_1 = require("./timertask");
class TaskFactory {
    static UncompletedTask(action) {
        return new task_1.Task(false, false, action);
    }
    static SuccessfulTask(action, result, timestamp, id, completedHistoryEventIndex) {
        return new task_1.Task(true, false, action, result, timestamp, id, undefined, completedHistoryEventIndex);
    }
    static FailedTask(action, reason, timestamp, id, completedHistoryEventIndex, exception) {
        return new task_1.Task(true, true, action, reason, timestamp, id, exception, completedHistoryEventIndex);
    }
    static CompletedTimerTask(action, timestamp, id, completedHistoryEventIndex) {
        return new timertask_1.TimerTask(true, action, timestamp, id, completedHistoryEventIndex);
    }
    static UncompletedTimerTask(action) {
        return new timertask_1.TimerTask(false, action);
    }
    static SuccessfulTaskSet(tasks, completionIndex, result) {
        return new taskset_1.TaskSet(true, false, tasks, completionIndex, result, undefined);
    }
    static FailedTaskSet(tasks, completionIndex, exception) {
        return new taskset_1.TaskSet(true, true, tasks, completionIndex, undefined, exception);
    }
    static UncompletedTaskSet(tasks) {
        return new taskset_1.TaskSet(false, false, tasks, undefined, undefined, undefined);
    }
}
exports.TaskFactory = TaskFactory;
