// durable-functions/lib/src/tasks/taskfilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const task_1 = require("./task");
const taskset_1 = require("./taskset");
class TaskFilter {
    static CompareFinishedTime(taskA, taskB) {
        if (taskA.completionIndex > taskB.completionIndex) {
            return 1;
        }
        if (taskA.completionIndex < taskB.completionIndex) {
            return -1;
        }
        return 0;
    }
    static isYieldable(task) {
        const taskBase = task;
        return (taskBase &&
            taskBase.isCompleted !== undefined &&
            taskBase.isFaulted !== undefined &&
            taskBase.yieldNewActions !== undefined);
    }
    static isSingleTask(task) {
        return task instanceof task_1.Task;
    }
    static isTaskSet(task) {
        return task instanceof taskset_1.TaskSet;
    }
    static isCompletedTask(task) {
        return task.isCompleted;
    }
    static isUncompletedTask(task) {
        return task.isCompleted === false;
    }
    static isSuccessfulTask(task) {
        return task.isCompleted === true && task.isFaulted === false;
    }
    static isFailedTask(task) {
        return task.isCompleted === true && task.isFaulted === true;
    }
}
exports.TaskFilter = TaskFilter;
